# Data Analyst - Take-Home Assignment 

## Know Data

Firstly, we need to load the python libraries and the dataset. For this exercise, I am using the data provided by capchase.

Import Libraries

In [2]:
import pandas as pd
import datetime as dt

Import Data

In [3]:
historical_invoices_df = pd.read_excel('../data/test_invoices_20210305.xlsx', index_col=0)

In [4]:
historical_invoices_df.head()

id payment_type    type       account  account_expiration  \
0  ID_13732           cc    sale  Account_4620                   5   
1  ID_13732           cc  settle  Account_4620                   5   
2  ID_14747           cc    sale  Account_4620                   5   
3  ID_14747           cc  settle  Account_4620                   5   
4  ID_19256           cc    sale  Account_1518                  74   

             time  amount    status response responsetext avs_results  \
0  20181207000604    1.00  complete  success     Approved           Z   
1  20181207023427    1.00  complete  success     ACCEPTED           Z   
2  20181207001202    2.00  complete  success     APPROVAL           Z   
3  20181207012743    2.00  complete  success     APPROVED           Z   
4  20181208022222    0.01  complete  success     Approved           0   

  csc_results     batch_id      first_name      last_name    company  
0           M          NaN  First_Name_987  Last_Name_918  Company_1  
1           M          1.0  First_Name_987  Last_Name_918  Company_1  
2           M          NaN  First_Name_987  Last_Name_918  Company_1  
3           M  337231373.0  First_Name_987  Last_Name_918  Company_1  
4           M          NaN             NaN            NaN        NaN

Before zooming into each field, let’s first take a bird’s eye view of the overall dataset characteristics.

In [5]:
# count of non-null values for each column and its data type.
historical_invoices_df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 41201 entries, 0 to 41200
Data columns (total 16 columns):
 #   Column              Non-Null Count  Dtype  
---  ------              --------------  -----  
 0   id                  41201 non-null  object 
 1   payment_type        41201 non-null  object 
 2   type                41201 non-null  object 
 3   account             41201 non-null  object 
 4   account_expiration  41201 non-null  int64  
 5   time                41201 non-null  int64  
 6   amount              41201 non-null  float64
 7   status              41201 non-null  object 
 8   response            41201 non-null  object 
 9   responsetext        41161 non-null  object 
 10  avs_results         39389 non-null  object 
 11  csc_results         40215 non-null  object 
 12  batch_id            19253 non-null  float64
 13  first_name          41186 non-null  object 
 14  last_name           41186 non-null  object 
 15  company             13 non-null     object 
dtypes: f

In [6]:
#basic statistics of each column. parameter “include = ‘all’”, it outputs the value count, unique count, the top-frequency value of the categorical variables and count, mean, standard deviation, min, max, and percentile of numeric variables
historical_invoices_df.describe(include='all')

id payment_type   type      account  account_expiration  \
count      41201        41201  41201        41201        41201.000000   
unique     21898            1      4         5474                 NaN   
top     ID_12872           cc   sale  Account_675                 NaN   
freq           3        41201  21439           64                 NaN   
mean         NaN          NaN    NaN          NaN           48.556661   
std          NaN          NaN    NaN          NaN           27.304527   
min          NaN          NaN    NaN          NaN            0.000000   
25%          NaN          NaN    NaN          NaN           23.000000   
50%          NaN          NaN    NaN          NaN           49.000000   
75%          NaN          NaN    NaN          NaN           72.000000   
max          NaN          NaN    NaN          NaN           93.000000   

                time        amount    status response responsetext  \
count   4.120100e+04  41201.000000     41201    41201        41161   
unique           NaN           NaN         4        2           27   
top              NaN           NaN  complete  success     APPROVED   
freq             NaN           NaN     38390    38494        19170   
mean    2.020094e+13    252.330717       NaN      NaN          NaN   
std     4.626674e+09    548.650148       NaN      NaN          NaN   
min     2.018121e+13  -2550.990000       NaN      NaN          NaN   
25%     2.020030e+13     50.000000       NaN      NaN          NaN   
50%     2.020071e+13    129.990000       NaN      NaN          NaN   
75%     2.020111e+13    199.990000       NaN      NaN          NaN   
max     2.021031e+13   7100.000000       NaN      NaN          NaN   

       avs_results csc_results      batch_id       first_name       last_name  \
count        39389       40215  1.925300e+04            41186           41186   
unique          11           3           NaN             1843            3707   
top              Z           M           NaN  First_Name_1631  Last_Name_3354   
freq         26537       39554           NaN              471             223   
mean           NaN         NaN  4.332791e+08              NaN             NaN   
std            NaN         NaN  2.249960e+07              NaN             NaN   
min            NaN         NaN  1.000000e+00              NaN             NaN   
25%            NaN         NaN  4.146724e+08              NaN             NaN   
50%            NaN         NaN  4.326441e+08              NaN             NaN   
75%            NaN         NaN  4.516896e+08              NaN             NaN   
max            NaN         NaN  4.704812e+08              NaN             NaN   

          company  
count          13  
unique          3  
top     Company_2  
freq            8  
mean          NaN  
std           NaN  
min           NaN  
25%           NaN  
50%           NaN  
75%           NaN  
max           NaN

Focus on identifying the number of missing values. isnull().sum()returns the number of missing values for each column.

In [7]:
historical_invoices_df.isnull().sum()

id                        0
payment_type              0
type                      0
account                   0
account_expiration        0
time                      0
amount                    0
status                    0
response                  0
responsetext             40
avs_results            1812
csc_results             986
batch_id              21948
first_name               15
last_name                15
company               41188
dtype: int64

In [8]:
missing_count = historical_invoices_df.isnull().sum()
value_count = historical_invoices_df.isnull().count()
missing_percentage = round(missing_count / value_count * 100, 2)
missing_percentage

id                     0.00
payment_type           0.00
type                   0.00
account                0.00
account_expiration     0.00
time                   0.00
amount                 0.00
status                 0.00
response               0.00
responsetext           0.10
avs_results            4.40
csc_results            2.39
batch_id              53.27
first_name             0.04
last_name              0.04
company               99.97
dtype: float64

After reviewing the data, I have reached the following conclusions:
1. The company column barely has values, 99.97% of the data is null, so I have decided to delete the column
2. To create the subscriptions, I don't think the batch_id column is necessary, understanding this as a batch of invoices. 
3. All columns with null values> 50% will be removed.

# 2. Data clean

Now we are going to remove all columns with null values> 50%

In [9]:
len(historical_invoices_df.columns)

16

In [10]:
threshold = 50

null_filter = missing_percentage > threshold

drop_cols = list(missing_percentage[null_filter].index)

drop_cols

['batch_id', 'company']

In [11]:
historical_invoices_df.drop(drop_cols, axis=1, inplace=True)

We standardize the column 'responsetext'

In [12]:
historical_invoices_df['responsetext'] = historical_invoices_df['responsetext'].str.lower()

Let's convert string timestamp columns into a proper date column using to_datetime function.

In [13]:
historical_invoices_df['time'] = pd.to_datetime(historical_invoices_df['time'], format='%Y%m%d%H%M%S') 
historical_invoices_df['time'].head()

0   2018-12-07 00:06:04
1   2018-12-07 02:34:27
2   2018-12-07 00:12:02
3   2018-12-07 01:27:43
4   2018-12-08 02:22:22
Name: time, dtype: datetime64[ns]

In [14]:
historical_invoices_df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 41201 entries, 0 to 41200
Data columns (total 14 columns):
 #   Column              Non-Null Count  Dtype         
---  ------              --------------  -----         
 0   id                  41201 non-null  object        
 1   payment_type        41201 non-null  object        
 2   type                41201 non-null  object        
 3   account             41201 non-null  object        
 4   account_expiration  41201 non-null  int64         
 5   time                41201 non-null  datetime64[ns]
 6   amount              41201 non-null  float64       
 7   status              41201 non-null  object        
 8   response            41201 non-null  object        
 9   responsetext        41161 non-null  object        
 10  avs_results         39389 non-null  object        
 11  csc_results         40215 non-null  object        
 12  first_name          41186 non-null  object        
 13  last_name           41186 non-null  object    

For the company name, we will create a new company column that will be formed by the first_name and last_name

In [16]:
historical_invoices_df['company'] = historical_invoices_df['first_name']  + ' ' + historical_invoices_df['last_name'] 

In [17]:
historical_invoices_df.isnull().sum()

id                       0
payment_type             0
type                     0
account                  0
account_expiration       0
time                     0
amount                   0
status                   0
response                 0
responsetext            40
avs_results           1812
csc_results            986
first_name              15
last_name               15
company                 15
dtype: int64

Solo me interesa analizar aquellas empresas con status complete, por ello aplicaré los siguiente filtros y podré determinar que valor de la columna responsetext podré utilizar para determinar el ciclo de facturación

In [19]:
status_filter = historical_invoices_df['status'] == 'complete'
response_filter = historical_invoices_df['response'] == 'success' 
response_filter_2 = historical_invoices_df['status'] == 'unknown' 
filter_full = (status_filter | response_filter_2)& response_filter 

In [20]:
invoices_df = historical_invoices_df[filter_full]

## 3. Data Process

Asumimos:
1. la columna account equivale a las cuentas de las empresas.
2. columna id hace relación al id de las facturas.
Finalmente vamos a agrupar por cuentas y facturas de esa cuenta. Una vez hecho esto procedemos a calcular el tiempo que transcurre entre las facturas consecutivas. 
Cuando una cuenta solo disponga de un invoice determinaré que el ciclo de facturación es anual. Es algo que podríamos revisar si nos dan un excel actualizado en el futuro.
Para definir qué facturas voy revisar filtraré por todas las invoices que tengan un status de approved, asumiendo que es cuando se aprueba y se envía la factura.



In [86]:
df_columns = ['account', 'id', 'payment_type', 'type','account_expiration', 'time', 
            'amount', 'status', 'response', 'responsetext', 'avs_results',
            'csc_results', 'first_name', 'last_name', 'company']

In [90]:
#invoices_df.sort_values('account').set_index('account').sort_values(by='time', ascending=True)

In [22]:
#invoice_group = invoices_df.groupby(['account']).count().reset_index()
#invoice_group.head()

account  id  payment_type  type  account_expiration  time  amount  \
0     Account_0   8             8     8                   8     8       8   
1     Account_1   8             8     8                   8     8       8   
2    Account_10  28            28    28                  28    28      28   
3   Account_100   4             4     4                   4     4       4   
4  Account_1000  10            10    10                  10    10      10   

   status  response  responsetext  avs_results  csc_results  first_name  \
0       8         8             8            4            4           8   
1       8         8             8            6            6           8   
2      28        28            28           28           28          28   
3       4         4             4            4            4           4   
4      10        10            10           10           10          10   

   last_name  company  
0          8        8  
1          8        8  
2         28       28  
3          4        4  
4         10       10

In [92]:
#invoice_group = invoices_df.groupby(df_columns).count().reset_index()
#invoice_group

account        id payment_type    type  account_expiration  \
0        Account_0  ID_14614           cc    sale                  21   
1        Account_0  ID_14614           cc  settle                  21   
2        Account_0  ID_19033           cc    sale                  21   
3        Account_0  ID_19033           cc  settle                  21   
4        Account_1  ID_19999           cc    sale                  49   
...            ...       ...          ...     ...                 ...   
36617  Account_997  ID_15450           cc  settle                  90   
36618  Account_997  ID_18730           cc    sale                  90   
36619  Account_997  ID_18730           cc  settle                  90   
36620  Account_999  ID_19667           cc    sale                  56   
36621  Account_999  ID_19667           cc  settle                  56   

                     time   amount    status response responsetext  \
0     2020-02-19 15:07:35   249.99  complete  success     approval   
1     2020-02-20 01:27:23   249.99  complete  success     approved   
2     2019-11-19 11:15:09   199.00  complete  success     approval   
3     2019-11-20 01:26:26   199.00  complete  success     approved   
4     2020-10-20 20:27:52    99.99  complete  success     approval   
...                   ...      ...       ...      ...          ...   
36617 2020-06-08 00:23:53   249.99  complete  success     approved   
36618 2020-06-18 18:31:09  1200.99  complete  success     approval   
36619 2020-06-19 00:28:21  1200.99  complete  success     approved   
36620 2020-04-02 18:14:32   149.99  complete  success     approval   
36621 2020-04-03 00:30:54   149.99  complete  success     approved   

      avs_results csc_results       first_name       last_name  \
0               Y           M   First_Name_728  Last_Name_3428   
1               Y           M   First_Name_728  Last_Name_3428   
2               Y           M   First_Name_728  Last_Name_3428   
3               Y           M   First_Name_728  Last_Name_3428   
4               Y           M  First_Name_1032  Last_Name_3670   
...           ...         ...              ...             ...   
36617           Z           M  First_Name_1524  Last_Name_1685   
36618           Z           M  First_Name_1524  Last_Name_1685   
36619           Z           M  First_Name_1524  Last_Name_1685   
36620           Y           M   First_Name_368  Last_Name_2910   
36621           Y           M   First_Name_368  Last_Name_2910   

                              company  
0       First_Name_728 Last_Name_3428  
1       First_Name_728 Last_Name_3428  
2       First_Name_728 Last_Name_3428  
3       First_Name_728 Last_Name_3428  
4      First_Name_1032 Last_Name_3670  
...                               ...  
36617  First_Name_1524 Last_Name_1685  
36618  First_Name_1524 Last_Name_1685  
36619  First_Name_1524 Last_Name_1685  
36620   First_Name_368 Last_Name_2910  
36621   First_Name_368 Last_Name_2910  

[36622 rows x 15 columns]

In [24]:
aproved_filter = historical_invoices_df['responsetext'] == 'approved' 
historical_invoices_df = historical_invoices_df[aproved_filter]
historical_invoices_df

id payment_type    type       account  account_expiration  \
0      ID_13732           cc    sale  Account_4620                   5   
3      ID_14747           cc  settle  Account_4620                   5   
4      ID_19256           cc    sale  Account_1518                  74   
7      ID_19313           cc  settle  Account_1518                  74   
9      ID_15498           cc  settle  Account_4620                   5   
...         ...          ...     ...           ...                 ...   
41106   ID_3390           cc  settle  Account_3162                  65   
41108  ID_18258           cc  settle  Account_1684                   6   
41110  ID_15065           cc  settle   Account_948                  32   
41112   ID_5926           cc  settle  Account_2748                  72   
41114   ID_5731           cc  settle  Account_2645                  40   

                     time   amount    status response responsetext  \
0     2018-12-07 00:06:04     1.00  complete  success     approved   
3     2018-12-07 01:27:43     2.00  complete  success     approved   
4     2018-12-08 02:22:22     0.01  complete  success     approved   
7     2018-12-09 01:21:24     0.02  complete  success     approved   
9     2019-01-15 01:31:03     1.00  complete  success     approved   
...                   ...      ...       ...      ...          ...   
41106 2021-03-04 01:24:15   199.99  complete  success     approved   
41108 2021-03-04 01:24:15   500.00  complete  success     approved   
41110 2021-03-04 01:24:15  1800.00  complete  success     approved   
41112 2021-03-04 01:24:15    99.99  complete  success     approved   
41114 2021-03-04 01:24:15   149.99  complete  success     approved   

      avs_results csc_results       first_name       last_name  \
0               Z           M   First_Name_987   Last_Name_918   
3               Z           M   First_Name_987   Last_Name_918   
4               0           M              NaN             NaN   
7             NaN           M              NaN             NaN   
9               Z           M   First_Name_987   Last_Name_918   
...           ...         ...              ...             ...   
41106           Z           M  First_Name_1250  Last_Name_2293   
41108           Y           M   First_Name_525  Last_Name_3098   
41110           Z           M   First_Name_241  Last_Name_1151   
41112           Y           M  First_Name_1510  Last_Name_3688   
41114           Z           M   First_Name_676  Last_Name_1481   

                              company  
0        First_Name_987 Last_Name_918  
3        First_Name_987 Last_Name_918  
4                                 NaN  
7                                 NaN  
9        First_Name_987 Last_Name_918  
...                               ...  
41106  First_Name_1250 Last_Name_2293  
41108   First_Name_525 Last_Name_3098  
41110   First_Name_241 Last_Name_1151  
41112  First_Name_1510 Last_Name_3688  
41114   First_Name_676 Last_Name_1481  

[19177 rows x 15 columns]

In [28]:
SELECTED_COLUMNS = ['account', 'company', 'time', 'amount']
df = historical_invoices_df[SELECTED_COLUMNS]
df

account                        company                time  \
0      Account_4620   First_Name_987 Last_Name_918 2018-12-07 00:06:04   
1      Account_4620   First_Name_987 Last_Name_918 2018-12-07 02:34:27   
2      Account_4620   First_Name_987 Last_Name_918 2018-12-07 00:12:02   
3      Account_4620   First_Name_987 Last_Name_918 2018-12-07 01:27:43   
4      Account_1518                            NaN 2018-12-08 02:22:22   
...             ...                            ...                 ...   
41196  Account_3921  First_Name_344 Last_Name_3600 2021-03-05 01:23:29   
41197  Account_4719  First_Name_111 Last_Name_1219 2021-03-04 23:41:09   
41198  Account_4719  First_Name_111 Last_Name_1219 2021-03-05 01:23:29   
41199  Account_4719  First_Name_111 Last_Name_1219 2021-03-05 00:21:50   
41200  Account_4719  First_Name_111 Last_Name_1219 2021-03-05 01:23:29   

        amount  
0         1.00  
1         1.00  
2         2.00  
3         2.00  
4         0.01  
...        ...  
41196  1100.00  
41197   199.99  
41198   199.99  
41199  2000.00  
41200  2000.00  

[41201 rows x 4 columns]

In [257]:
accounts = list(df['account'].unique())

In [258]:
results = []
for account in accounts:
    df_2 = df[df['account']== account].sort_values(by='time', ascending=True)
    df_2['difference'] = df_2['time'].diff()
    mean_time = df_2['difference'].mean()
    
    results.append((account,mean_time,))
    

In [29]:
FREQUENCY = {
    'weekly': {
        'name': 'weekly',
        'min_days': pd.Timedelta(days=1),
        'max_days': pd.Timedelta(days=7),
        'sum_time': pd.Timedelta(days=7)
    },
    'monthly': {
        'name': 'monthly',
        'min_days': pd.Timedelta(days=7),
        'max_days': pd.Timedelta(days=31),
        'sum_time': pd.Timedelta(days=30)
    },
    'quarterly': {
        'name': 'quaterly',
        'min_days': pd.Timedelta(days=30),
        'max_days': pd.Timedelta(days=92),
        'sum_time': pd.Timedelta(days=90)
    },
    'semesterly': {
        'name': 'semesterly',
        'min_days': pd.Timedelta(days=90),
        'max_days': pd.Timedelta(days=181),
        'sum_time': pd.Timedelta(days=181)
    },
    'annualy': {
        'name': 'annualy',
        'min_days': pd.Timedelta(days=181),
        'max_days': pd.Timedelta(days=365),
        'sum_time': pd.Timedelta(days=365)
    }
}


In [303]:
def get_freq(row_difference_time): 
    #print(row_difference_time)
    if FREQUENCY['weekly']['min_days'] < row_difference_time <= FREQUENCY['weekly']['max_days']:
        return (FREQUENCY['weekly']['name'], row_difference_time + FREQUENCY['weekly']['sum_time'])
    elif FREQUENCY['monthly']['min_days'] < row_difference_time <= FREQUENCY['monthly']['max_days']:
        return (FREQUENCY['monthly']['name'], row_difference_time + FREQUENCY['monthly']['sum_time'])
    elif FREQUENCY['quarterly']['min_days'] < row_difference_time <= FREQUENCY['quarterly']['max_days']:
        return (FREQUENCY['quarterly']['name'], row_difference_time + FREQUENCY['quarterly']['sum_time'])
    elif FREQUENCY['semesterly']['min_days'] < row_difference_time <= FREQUENCY['semesterly']['max_days']:
        return (FREQUENCY['semesterly']['name'], row_difference_time + FREQUENCY['semesterly']['sum_time'])
    else:
        return (FREQUENCY['annualy']['name'], row_difference_time + FREQUENCY['annualy']['sum_time'])

In [434]:
df_total = pd.DataFrame(results, columns = ['account','difference_time_mean']).merge( 
                    df[['time','account', 'company', 'amount']].groupby('account').max().reset_index().rename(columns={'time':'last_time'}), how='left')
df_total

account    difference_time_mean           last_time  \
0     Account_4620 68 days 05:04:13.800000 2019-11-13 01:27:13   
1     Account_1518         0 days 22:59:02 2018-12-09 01:21:24   
2      Account_859        17 days 05:28:36 2019-11-25 16:28:45   
3     Account_1666                     NaT 2019-11-17 01:21:44   
4     Account_5114        30 days 12:03:00 2020-01-17 01:27:44   
...            ...                     ...                 ...   
4931  Account_4357                     NaT 2021-03-04 01:24:15   
4932  Account_4514                     NaT 2021-03-04 01:24:15   
4933  Account_4814                     NaT 2021-03-04 01:24:15   
4934  Account_2748                     NaT 2021-03-04 01:24:15   
4935  Account_2645                     NaT 2021-03-04 01:24:15   

                             company   amount  
0       First_Name_987 Last_Name_918     2.00  
1                                NaN     0.02  
2                                NaN     0.03  
3     First_Name_1563 Last_Name_1132  1140.99  
4      First_Name_409 Last_Name_1236    89.99  
...                              ...      ...  
4931  First_Name_1326 Last_Name_2571   299.99  
4932   First_Name_1836 Last_Name_455  1100.00  
4933  First_Name_1524 Last_Name_2725   149.99  
4934  First_Name_1510 Last_Name_3688    99.99  
4935   First_Name_676 Last_Name_1481   149.99  

[4936 rows x 5 columns]

In [435]:
df_total['periodValue']= df_total['difference_time_mean'].apply(get_freq)

In [436]:
df_total[['split_period', 'next_time']] = pd.DataFrame(df_total['periodValue'].tolist(), index=df_total.index)  

In [437]:
def get_difference_time_null(df):
    if pd.isnull(df['difference_time_mean']):
        df['next_time'] = pd.Timedelta(days=365)
    return df

In [438]:
df_total = df_total.apply(lambda x: get_difference_time_null(x), axis=1)

In [445]:
df_total.head(20)

account       difference_time_mean           last_time  \
0   Account_4620    68 days 05:04:13.800000 2019-11-13 01:27:13   
1   Account_1518            0 days 22:59:02 2018-12-09 01:21:24   
2    Account_859           17 days 05:28:36 2019-11-25 16:28:45   
3   Account_1666                        NaT 2019-11-17 01:21:44   
4   Account_5114           30 days 12:03:00 2020-01-17 01:27:44   
5    Account_318    32 days 23:47:34.500000 2020-03-28 00:32:02   
6   Account_2960           30 days 00:06:57 2019-12-17 01:28:41   
7   Account_4801 30 days 10:33:52.666666666 2020-08-17 00:26:38   
8   Account_1409    30 days 08:53:40.250000 2020-07-17 00:31:06   
9   Account_1199 30 days 15:59:43.333333333 2020-02-17 01:20:54   
10  Account_3304 30 days 12:48:03.133333333 2021-02-17 01:22:31   
11  Account_4054 30 days 12:48:03.066666666 2021-02-17 01:22:31   
12  Account_3358 30 days 12:48:03.066666666 2021-02-17 01:22:31   
13  Account_3136    30 days 05:47:36.500000 2020-03-17 00:32:11   
14  Account_3806           30 days 15:59:43 2020-02-17 01:20:54   
15  Account_4954 35 days 11:51:19.666666666 2020-06-17 00:29:43   
16  Account_2546    30 days 05:47:36.500000 2020-03-17 00:32:11   
17  Account_2893           30 days 00:06:56 2019-12-17 01:28:41   
18  Account_3413 30 days 12:48:03.066666666 2021-02-17 01:22:31   
19  Account_2012                        NaT 2019-11-17 01:21:45   

                           company   amount split_period  \
0     First_Name_987 Last_Name_918     2.00     quaterly   
1                              NaN     0.02      annualy   
2                              NaN     0.03      monthly   
3   First_Name_1563 Last_Name_1132  1140.99      annualy   
4    First_Name_409 Last_Name_1236    89.99      monthly   
5    First_Name_165 Last_Name_3196    89.99     quaterly   
6   First_Name_1618 Last_Name_2858    79.99      monthly   
7    First_Name_395 Last_Name_3360    15.00      monthly   
8   First_Name_1408 Last_Name_2470    99.99      monthly   
9     First_Name_420 Last_Name_918    79.99      monthly   
10  First_Name_1055 Last_Name_3705    79.99      monthly   
11    First_Name_430 Last_Name_914   149.99      monthly   
12  First_Name_1563 Last_Name_1336   159.99      monthly   
13   First_Name_129 Last_Name_2360   149.99      monthly   
14   First_Name_340 Last_Name_1315   149.99      monthly   
15     First_Name_513 Last_Name_23    30.00     quaterly   
16   First_Name_964 Last_Name_2265   199.99      monthly   
17    First_Name_859 Last_Name_793   199.99      monthly   
18   First_Name_1718 Last_Name_295    49.99      monthly   
19   First_Name_576 Last_Name_3495   149.99      annualy   

                     next_time               nextPaymentDate  
0     158 days 05:04:13.800000 2020-04-19 06:31:26.800000000  
1            365 days 22:59:02 2019-12-10 00:20:26.000000000  
2             47 days 05:28:36 2020-01-11 21:57:21.000000000  
3            365 days 00:00:00 2020-11-16 01:21:44.000000000  
4             60 days 12:03:00 2020-03-17 13:30:44.000000000  
5     122 days 23:47:34.500000 2020-07-29 00:19:36.500000000  
6             60 days 00:06:57 2020-02-15 01:35:38.000000000  
7   60 days 10:33:52.666666666 2020-10-16 11:00:30.666666666  
8      60 days 08:53:40.250000 2020-09-15 09:24:46.250000000  
9   60 days 15:59:43.333333333 2020-04-17 17:20:37.333333333  
10  60 days 12:48:03.133333333 2021-04-18 14:10:34.133333333  
11  60 days 12:48:03.066666666 2021-04-18 14:10:34.066666666  
12  60 days 12:48:03.066666666 2021-04-18 14:10:34.066666666  
13     60 days 05:47:36.500000 2020-05-16 06:19:47.500000000  
14            60 days 15:59:43 2020-04-17 17:20:37.000000000  
15 125 days 11:51:19.666666666 2020-10-20 12:21:02.666666666  
16     60 days 05:47:36.500000 2020-05-16 06:19:47.500000000  
17            60 days 00:06:56 2020-02-15 01:35:37.000000000  
18  60 days 12:48:03.066666666 2021-04-18 14:10:34.066666666  
19           365 days 00:00:00 2020-11-16 01:21:45.000000000

In [440]:
df_total.isnull().sum()

account                    0
difference_time_mean    1468
last_time                  0
company                    3
amount                     0
periodValue                0
split_period               0
next_time                  0
dtype: int64

In [441]:
df_total.drop(columns='periodValue', inplace=True)

In [442]:
df_total.rename(columns={'split_period':'periodValue'})

account    difference_time_mean           last_time  \
0     Account_4620 68 days 05:04:13.800000 2019-11-13 01:27:13   
1     Account_1518         0 days 22:59:02 2018-12-09 01:21:24   
2      Account_859        17 days 05:28:36 2019-11-25 16:28:45   
3     Account_1666                     NaT 2019-11-17 01:21:44   
4     Account_5114        30 days 12:03:00 2020-01-17 01:27:44   
...            ...                     ...                 ...   
4931  Account_4357                     NaT 2021-03-04 01:24:15   
4932  Account_4514                     NaT 2021-03-04 01:24:15   
4933  Account_4814                     NaT 2021-03-04 01:24:15   
4934  Account_2748                     NaT 2021-03-04 01:24:15   
4935  Account_2645                     NaT 2021-03-04 01:24:15   

                             company   amount periodValue  \
0       First_Name_987 Last_Name_918     2.00    quaterly   
1                                NaN     0.02     annualy   
2                                NaN     0.03     monthly   
3     First_Name_1563 Last_Name_1132  1140.99     annualy   
4      First_Name_409 Last_Name_1236    89.99     monthly   
...                              ...      ...         ...   
4931  First_Name_1326 Last_Name_2571   299.99     annualy   
4932   First_Name_1836 Last_Name_455  1100.00     annualy   
4933  First_Name_1524 Last_Name_2725   149.99     annualy   
4934  First_Name_1510 Last_Name_3688    99.99     annualy   
4935   First_Name_676 Last_Name_1481   149.99     annualy   

                    next_time  
0    158 days 05:04:13.800000  
1           365 days 22:59:02  
2            47 days 05:28:36  
3           365 days 00:00:00  
4            60 days 12:03:00  
...                       ...  
4931        365 days 00:00:00  
4932        365 days 00:00:00  
4933        365 days 00:00:00  
4934        365 days 00:00:00  
4935        365 days 00:00:00  

[4936 rows x 7 columns]

In [443]:
df_total['nextPaymentDate'] = df_total['next_time'] + df_total['last_time']

In [444]:
df_total

account    difference_time_mean           last_time  \
0     Account_4620 68 days 05:04:13.800000 2019-11-13 01:27:13   
1     Account_1518         0 days 22:59:02 2018-12-09 01:21:24   
2      Account_859        17 days 05:28:36 2019-11-25 16:28:45   
3     Account_1666                     NaT 2019-11-17 01:21:44   
4     Account_5114        30 days 12:03:00 2020-01-17 01:27:44   
...            ...                     ...                 ...   
4931  Account_4357                     NaT 2021-03-04 01:24:15   
4932  Account_4514                     NaT 2021-03-04 01:24:15   
4933  Account_4814                     NaT 2021-03-04 01:24:15   
4934  Account_2748                     NaT 2021-03-04 01:24:15   
4935  Account_2645                     NaT 2021-03-04 01:24:15   

                             company   amount split_period  \
0       First_Name_987 Last_Name_918     2.00     quaterly   
1                                NaN     0.02      annualy   
2                                NaN     0.03      monthly   
3     First_Name_1563 Last_Name_1132  1140.99      annualy   
4      First_Name_409 Last_Name_1236    89.99      monthly   
...                              ...      ...          ...   
4931  First_Name_1326 Last_Name_2571   299.99      annualy   
4932   First_Name_1836 Last_Name_455  1100.00      annualy   
4933  First_Name_1524 Last_Name_2725   149.99      annualy   
4934  First_Name_1510 Last_Name_3688    99.99      annualy   
4935   First_Name_676 Last_Name_1481   149.99      annualy   

                    next_time         nextPaymentDate  
0    158 days 05:04:13.800000 2020-04-19 06:31:26.800  
1           365 days 22:59:02 2019-12-10 00:20:26.000  
2            47 days 05:28:36 2020-01-11 21:57:21.000  
3           365 days 00:00:00 2020-11-16 01:21:44.000  
4            60 days 12:03:00 2020-03-17 13:30:44.000  
...                       ...                     ...  
4931        365 days 00:00:00 2022-03-04 01:24:15.000  
4932        365 days 00:00:00 2022-03-04 01:24:15.000  
4933        365 days 00:00:00 2022-03-04 01:24:15.000  
4934        365 days 00:00:00 2022-03-04 01:24:15.000  
4935        365 days 00:00:00 2022-03-04 01:24:15.000  

[4936 rows x 8 columns]